processor/src/main/java/com/linecorp/decaton/processor/runtime/TracingProvider.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.decaton.processor.runtime;
+
+import java.io.Closeable;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+import com.linecorp.decaton.client.KafkaProducerSupplier;
+
+/**
+ * Interface for distributed tracing implementations that track traces via Kafka {@link ConsumerRecord}s
+ * (typically through Kafka headers) e.g. Zipkin, OpenTracing, Kamon.
+ * Note that this will usually need to be paired with a suitable {@link KafkaProducerSupplier} on the producer
+ * side to write those Kafka headers.
+ * In particular, if you want retries to be shown as part of the same (distributed) trace, make sure to supply
+ * such a {@link KafkaProducerSupplier} in your {@link RetryConfig}.
+ */
+public interface TracingProvider {
```

Comment:
I went through the same thought process myself. If we want to pass a customisable `TracingContext` down into the processing implementation, then it would have to be carried somewhere (e.g. `ProcessingContext`) and would then have to be exposed as a type parameter on that, and on the `DecatonProcessor` as well - it seems too intrusive for something that's just meant to be tracing. And we would still need to do some work to make the trace context "current" on the thread that's executing the processor, so we'd either need to add more to this interface (or at least provide some non-obvious guarantees about which thread will call `onStage` with which value) or the user would have to write boilerplate in their processor.

I believe all the major tracing implementations use a thread-local context - at least Zipkin, OpenTracing, New Relic and Kamon do. So I think it's better to acknowledge that and design an interface that will work well for the tracing libraries that we're likely to support. Unless there's actually a non-thread-local-based tracing library that we want to implement support for?

Reply:
> So I think it's better to acknowledge that and design an interface that will work well for the tracing libraries that we're likely to support. 

Yeah, that's really a good point.
However, my concern is not only about applicability of this interface for any possible instrumentation libraries, but also maintainability of the protocol between decaton core and drivers.
By making this interface to implicitly relying on Thread context, that would makes decaton core impl need to take care of it when considering it's internal threading model. As you did in propagateXX method, for every place that decaton switches the thread it needs to ensure calling a method to propagate context to the next thread which sounds a fragile implication that can be easily broken when decaton's internal architecture changes (even if tests are supposed to catch it).
I'm also not sure if propagation always happen through Runnable interface. e.g, previous decaton implementation used a queue with a task type which will be received by a worker (a already running Runnable) on the other side.
So I think it's the best if we can make the list of information that are exchanged between decaton core and tracing driver always explicit on the interface, rather than expecting the thread local to hold particular information at a place.

> would then have to be exposed as a type parameter on that, and on the DecatonProcessor as well - it seems too intrusive for something that's just meant to be tracing. 

and this is true. So I think a good compromise is to let decaton processor code (user code) to extract their expected tracing type out of interface type by casting, if necessary.
While I don't prefer to introduce such dynamic casting in defined interface as possible as we can, I think it is essentially not much different from extracting context from the thread local. Both relies on a particular information that is not certain if it really exists (expected type vs potentially missing thread local field) from the processor's limited point of view. I mean - that's what they cannot be benefited by static checking.

``java
tc = (BraveTracingContext) context.tracingContext(); // or context.<BraveTracingContext>tracingContextAs() ?
vs
Tracing.getCurrentContext()
```

Btw as long as the instrumentation library uses thread local for holding a context, I think users won't need to access specific TracingProvider to instrument their synchronous processing parts? While I'm against to use thread local for bringing around the context inside the decaton core, I think it's good to setup (inside the driver, at `onStage`) thread local around calling `process` so users can just use the library in regular way. This should be fine because it's a protocol between each specific drivers and user that selects to use it and decaton core doesn't care about it (it just calls onStage).

> still need to do some work to make the trace context "current" on the thread that's executing the processor, so we'd either need to add more to this interface (or at least provide some non-obvious guarantees about which thread will call onStage with which value) 

This sounds fine though? "onStage(X) is called at place where X starts, by a thread that actually executes X" sounds like a reasonable contract to follow. We just don't want to let drivers to assume that the thread executes X and thread executes Y are contagious and can always "propagate" tracing context through thread local.

Reply:
Ah, I see. Yeah, that makes sense. I've reworked the interface with a similar design to what you suggest; please take a look and see what you think.